gamemodes/terrortown/gamemode/shared/sh_marker_vision_object.lua

```diff
@@ -0,0 +1,244 @@
+end
+
+---
+-- Updates the relations of this marker vision element.
+-- @note This does not sync it to the client, call @{MARKER_VISION_OBJECT:SyncToClients} to sync to client.
+-- @param number|string|Player owner The owner of the wallhack that takes their ownership with them on
+-- team change; can also be a team (string) or role (number) if it shouldn't be bound to a player
+-- @param number visibleFor Visibility setting: `VISIBLE_FOR_PLAYER`, `VISIBLE_FOR_ROLE`, `VISIBLE_FOR_TEAM`,
+-- `VISIBLE_FOR_ALL`
+-- @realm shared
+function MARKER_VISION_OBJECT:UpdateRelations(owner, visibleFor)
+	-- this is done this way to support an UpdateRelations call on Role/Team change without
+	-- the need to pass these values again
+	self.data.owner = owner or self.data.owner
+	self.data.visibleFor = visibleFor or self.data.visibleFor
+
+	if SERVER then
+		if self.data.visibleFor == VISIBLE_FOR_PLAYER then
+			self.receiverList = {self.data.owner}
+		elseif self.data.visibleFor == VISIBLE_FOR_ROLE then
+			if IsPlayer(self.data.owner) then
+				self.receiverList = GetRoleFilter(self.data.owner:GetSubRole(), false)
+			else
+				-- handle static role
+				self.receiverList = GetRoleFilter(self.data.owner, false)
+			end
+		elseif self.data.visibleFor == VISIBLE_FOR_TEAM then
+			if IsPlayer(self.data.owner) then
+				self.receiverList = GetTeamFilter(self.data.owner:GetTeam(), false, true)
+			else
+				-- handle static team
+				self.receiverList = GetTeamFilter(self.data.owner, false, true)
+			end
+		else
+			self.receiverList = nil
+		end
+
+		-- note: when VISIBLE_FOR_ALL is used, the receiver will be nil and
+		-- therefore SendStream uses net.Broadcast
+	end
+end
+
+---
+-- Checks if the entity is valid and the identifier is set.
+-- @return boolean Returns true if valid
+-- @realm shared
+function MARKER_VISION_OBJECT:IsValid()
+	if not IsValid(self.data.ent) then
+		return false
+	end
+
+	if not self.data.identifier or self.data.identifier == "" then
+		return false
+	end
+
+	return true
+end
+
+if SERVER then
+	---
+	-- Syncs the marker vision element to the client that should receive it.
+	-- @note The receipients should be set first with @{MARKER_VISION_OBJECT:UpdateRelations}.
+	-- @note If the marker vision element was synced already, it is updated on the client.
+	-- @param[opt] table receiverListOverwrite A table of players that should receive the update
+	-- if not the automatic receipient selection should be used.
+	-- @param[opt] any passThroughData Data that should be sent with this call as well
+	-- @realm server
+	function MARKER_VISION_OBJECT:SyncToClients(receiverListOverwrite, passThroughData)
+		if not self.data.owner then return end
+
+		-- if data was previously set, it should be removed first on the client
+		if self.lastReceiverList then
+			net.Start("ttt2_marker_vision_entity_removed")
+			net.WriteEntity(self.data.ent)
+			net.WriteString(self.data.identifier)
+			net.Send(self.lastReceiverList)
+		end
```

Comment:
It is the easiest way to handle two things:
- a player that lost access to a markerVision element 
- a marker vision element that changed color as one has to remove and readd the mark client-side to achieve a color change

Reply:
I want to add here that a more in depth update system would probably work as well, but this would be way more complex code wise. So I chose this simple approach as it does the job without much complexity

Reply:
You are misunderstanding me again. I am asking those questions not because i necessarily want you to change this but rather as something that i asked myself when reading this and thus i noted that you might want to elaborate on the comment. 
These cases that you just listed are a perfect example for a comment. Something i was not aware of when reading the code. This makes it so much easier to refactor this later, because i now know why this was done here.